src/k_smallest.rs

```diff
@@ -1,20 +1,211 @@
+    // while reusing the storage
+    len: usize,
+}
+
+impl<T, C> MaxHeap<T, C>
+where
+    C: Fn(&T, &T) -> Ordering,
+{
+    fn from_iter<I>(k: usize, comparator: C, mut iter: I) -> Self
+    where
+        I: Iterator<Item = T>,
+    {
+        let storage: Vec<T> = iter.by_ref().take(k).collect();
+
+        let mut heap = Self {
+            len: storage.len(),
+            storage,
+            comparator,
+        };
+        // Filling up the storage and only afterwards rearranging to form a valid heap is slightly more efficient
+        // (But only by a factor of lg(k) and I'd love to hear of a usecase where that matters)
+        heap.heapify();
+
+        if k == heap.storage.len() {
+            // Nothing else needs done if we didn't fill the storage in the first place
+            // Also avoids unexpected behaviour with restartable iterators
+            for val in iter {
+                heap.push_pop(val);
+            }
+        }
+        heap
+    }
+
+    /// Retrieves the element at the given index.
+    /// Returning an Option means that bounds checking is subsumed in checking the
+    /// element ordering.
+    fn get(&self, index: usize) -> Option<&T> {
+        if index < self.len {
+            self.storage.get(index)
+        } else {
+            None
+        }
+    }
+
+    /// Sift the element currently at `origin` **away** from the root until it is properly ordered
+    fn sift_down(&mut self, origin: usize) {
+        fn children_of(n: usize) -> (usize, usize) {
+            (2 * n, 2 * n + 1)
+        }
+
+        let (left_idx, right_idx) = children_of(origin);
+        let (left_item, right_item) = (self.get(left_idx), self.get(right_idx));
+
+        if left_item.is_none() { // the left is the earlier child, so if it doesn't exist there's nothing to swap with
+            return;
+        }
+
+        let cmp = self
+            .compare(left_item, right_item)
+            .unwrap_or(Ordering::Greater); // The right item may not exist, so default to picking the left, i.e. lower
+
+        let replacement_idx = match cmp {
+            Ordering::Greater | Ordering::Equal => left_idx,
+            Ordering::Less => right_idx,
+        };
 
-    let mut heap = iter.by_ref().take(k).collect::<BinaryHeap<_>>();
+        let (original_item, replacement_item) = (self.get(origin), self.get(replacement_idx));
 
-    iter.for_each(|i| {
-        debug_assert_eq!(heap.len(), k);
-        // Equivalent to heap.push(min(i, heap.pop())) but more efficient.
-        // This should be done with a single `.peek_mut().unwrap()` but
-        //  `PeekMut` sifts-down unconditionally on Rust 1.46.0 and prior.
-        if *heap.peek().unwrap() > i {
-            *heap.peek_mut().unwrap() = i;
+        let cmp = self.compare(original_item, replacement_item);
+        if Some(Ordering::Less) == cmp {
+            self.storage.swap(origin, replacement_idx);
+            self.sift_down(replacement_idx);
```

Comment:
Nitpick (but I wouldn't be too strict about this if we do not run into actual problems): Should we write `sift_down` iteratively, avoiding the stack memory consumption of `log(n)`?

Reply:
My expectation would be that the compiler [already does that](https://llvm.org/docs/Passes.html#tailcallelim-tail-call-elimination), so it takes `O(1)` stack space regardless. However, it might be worth writing it explicitly as a loop regardless, so the space consumption is clearer to the reader.